Add basic <audio>, <video> and <source> support for HtmlRenderer

iglosiggio opened the pull request:
This patch adds basic support for the <audio> and <video> tags emiting a new link for each available <source> (the MediaElement src is included in the available sources).

A feed that can be used to test parts of this patch is the [Factorio Blog](https://www.factorio.com/blog/rss).

Minoru commented on src/htmlrenderer.cpp:
```diff
@@ -730,6 +839,36 @@ void HtmlRenderer::render(std::istream& input,
+				} else {
+					LOG(Level::WARN,
+						"HtmlRenderer::render: found "
+						"video "
```
I believe this should say "audio".

iglosiggio commented:
Yup, it should say audio.

Should the tests consider only “good cases”? (well-formed usage of the
tags).

Minoru commented:
> Should the tests consider only “good cases”? (well-formed usage of the tags).

Ideally they would also check all the error cases that you've coded for, but if that's too much—please at least check the basic functionality.

iglosiggio commented:
@Minoru i've added tests similar to those for the <img> tag. I'm not testing unclosed <video>/<audio> tags but i'm testing tags w/o valid sources. I've also changed the way that the links are formatted.

I hope you like the new changes :)

iglosiggio commented on src/htmlrenderer.cpp:
```diff
@@ -91,11 +94,16 @@ void HtmlRenderer::render(std::istream& input,
 	unsigned int image_count = 0;
+	unsigned int video_count = 0;
+	unsigned int audio_count = 0;
+	unsigned int source_count = 0;
```
This isn't used. Should i remove it?

Minoru replied:
Yeah, I guess you should. It's easy to see that \<video>/\<audio> lacks sources (by inspecting the raw RSS/Atom), and there's nothing we can do about those missing tags inside Newsboat; so let's get rid of useless code.

iglosiggio commented on src/htmlrenderer.cpp:
```diff
@@ -482,6 +490,102 @@ void HtmlRenderer::render(std::istream& input,
+							_("link"),
+							link_num));
+				}
+				source_count++;
```
The increment should happen only if the `<source>` tag has a valid source.

Minoru replied:
It'd be nice to have a test for that :) Eventually we'll rewrite this in Rust (#89), so we better have a test suite that catches the likely bugs.

Minoru replied:
This wasn't addressed, was it? We still don't have a test that shows that empty tags don't increase the count.

iglosiggio replied:
I could extend the test for tags without valid sources to include tags with some valid sources. I'm sure that the code does the right thing (because `HtmlRenderer::add_media_link` does nothing if the link is an empty string).

Minoru replied:
Such a test would be great; thanks!

iglosiggio commented on src/htmlrenderer.cpp:
```diff
@@ -636,6 +740,7 @@ void HtmlRenderer::render(std::istream& input,
 			case HtmlTag::ITUNESHACK:
 			case HtmlTag::IMG:
 			case HtmlTag::HR:
+			case HtmlTag::SOURCE:
```
The [specification ](https://html.spec.whatwg.org/multipage/embedded-content.html#the-source-element) says that `<source>` tags don't close.

iglosiggio commented on src/htmlrenderer.cpp:
```diff
@@ -764,7 +899,7 @@ void HtmlRenderer::render(std::istream& input,
 				}
-			} else if (inside_script || inside_style) {
+			} else if (inside_script || inside_style || inside_video || inside_audio) {
 				// skip scripts and CSS styles
```
... and fallback text for videos

iglosiggio commented:
I've added a few comments but i will wait for your review before fixing the things i pointed out.

Minoru commented:
> I'm not testing unclosed \<video>/\<audio> tags

Any particular reason not to test those cases?

> I've also changed the way that the links are formatted.

That's a great idea, I like it! Even though it makes video/audio links look different from images, it's still better than making users think there are multiple videos where in fact there's multiple versions of the same.

On a side note: please don't force-push. GitHub doesn't give me all the diffs between old and new versions of commits, and I have to re-review everything from the very beginning. A better approach is to create fixup commits (`git commit --fixup=COMMIT_TO_FIX`), and to squash them (`git rebase --interactive --autosquash master`) after I did the final review.

Fixup commits might not work that well for large PRs that introduce a lot of changes; in that case, you can just make ordinary commits that address the review comments. The history will be a bit dirtier, but at least you won't risk introducing bugs while resolving merge conflicts.

Also, since Markdown allows to embed HTML, you have to escape it to display tags: when you write `I've added tests similar to <img>`, the `<img>` part is not shown because it's interpreted as embedded HTML. Put backslash in front of an opening tag to fix that: `I've added tests similar to \<img>`. I already did that to your last comment.

Minoru commented on src/htmlrenderer.cpp:
```diff
@@ -482,6 +490,102 @@ void HtmlRenderer::render(std::istream& input,
+							video_count,
+							_("link"),
+							link_num));
+					source_count++;
```
The contents of this `if` is repeated further down, in `case HtmlTag::SOURCE`. Would it be possible to move this into a function or a private method, and call that from both places? It can even be generalized to be used with `HtmlTag::VIDEO`, unless it hurts readability.

iglosiggio replied:
I've added

Minoru replied:
It looks excellent, thank you!

Minoru commented on src/htmlrenderer.cpp:
```diff
@@ -482,6 +490,102 @@ void HtmlRenderer::render(std::istream& input,
+					videourl = "";
+				}
+				video_count++;
+				if (videourl.length() > 0) {
```
I guess this was just copied from `HtmlTag::IMG` handler, but it's more idiomatic to write `if (!videourl.empty())`.

iglosiggio commented:
> > I'm not testing unclosed <video>/<audio> tags
> Any particular reason not to test those cases?
They should be incredibly rare. And i really don't know what would be the expected outcome.

> On a side note: please don't force-push. GitHub doesn't give me all the diffs between old and new versions of commits, and I have to re-review everything from the very beginning. A better approach is to create fixup commits (`git commit --fixup=COMMIT_TO_FIX`), and to squash them (`git rebase --interactive --autosquash master`) after I did the final review.
Thanks for the tip!

> Also, since Markdown allows to embed HTML, you have to escape it to display tags
Whoops, i didn't noticed :sweat_smile:

Minoru commented:
> [Unclosed \<audio>/\<video> tags] should be incredibly rare. And i really don't know what would be the expected outcome.

[Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle) state that Newsboat should still be prepared for this, however unlikely.

I believe the standard practice is to implicitly close the previous tag when the next one is found.

iglosiggio commented:
> I believe the standard practice is to implicitly close the previous tag when the next one is found.

Oh! Then i would implement that :)

iglosiggio commented:
Whooops, it seems like i made a mess with the fixups

Minoru commented on src/htmlrenderer.cpp:
```diff
@@ -482,6 +489,69 @@ void HtmlRenderer::render(std::istream& input,
 			}
+
+			case HtmlTag::VIDEO: {
+				if (inside_video || inside_audio) {
```
Since these variables aren't booleans, it's better to explicitly check for equality: `if (inside_video != 0 || inside_audio != 0)`. (I'd change it in the \<img> code as well, but that's up to you.)

iglosiggio replied:
Isn't changing them to booleans a better solution? They cannot be nested and I've added the implicit-closing code.

Minoru replied:
Ah, that'd be even better!

Minoru commented on src/htmlrenderer.cpp:
```diff
@@ -821,6 +904,23 @@ void HtmlRenderer::render(std::istream& input,
+	}
+
+	std::string type_str = type2str(type);
+	int link_num = add_link(links,
```
This should be an `unsigned int`, since that's what `add_link` returns.

iglosiggio replied:
Oh, sorry, I didn't check the return type and copied the declaration in HtmlRenderer::render

Minoru commented on test/htmlrenderer.cpp:
```diff
@@ -1204,3 +1204,144 @@ TEST_CASE("When rendeing text, HtmlRenderer strips leading whitespace",
+	REQUIRE(lines[1] == p(LineType::wrappable, ""));
+	REQUIRE(lines[2] == p(LineType::wrappable, "Links: "));
+	REQUIRE(links.size() == 5);
+	REQUIRE(lines[0] == p(LineType::wrappable, "[video 1 (link #1)]"
```
It's a bit confusing that this line is at the end—please move it up, before `lines[1]` check.

Also, please check the contents of the `links`, to ensure the URLs were assigned the correct numbers.

iglosiggio replied:
Okay!

Minoru commented:
> I've fixed the squash conflicts (the resulting tree is the same and the codes compiles and passes the test in every commit). If you don't have any corrections over the current code i will force-push the corrected history

Thanks, but it appears that you've also rebased some of the commits I've already merged to master. If you force-push that, we'll have two copies of those commits: the ones we already have, plus the ones you rebased. Let's avoid that.

I think this happened because you used an older commit as a base. I believe you should've rebased onto https://github.com/newsboat/newsboat/commit/f80387f9c4a496258d6839bc472c8a0dc6b9e58f, i.e. the original commit from which you started basic-audio-video-support branch.

I've found a couple more nitpicks, but once they're fixed, I'll finally merge this. Thanks for patiently addressing all my comments, @iglosiggio!

iglosiggio commented:
Done! [i've also updated the branch with the clean history, now rebased correctly](https://github.com/iglosiggio/newsboat/tree/basic-audio-video-support-manually-squashed)

Minoru commented:
The new commits and the rebase look great; thanks! The only remaining thing seem to be [this comment](https://github.com/newsboat/newsboat/pull/729#discussion_r370967908), right?

iglosiggio commented:
That comment is no longer relevant as `source_count` was deleted :)

iglosiggio commented:
I've force-pushed the corrected history, there aren't any changes: https://github.com/newsboat/newsboat/compare/34acebb..3fe54f2

iglosiggio commented:
I seems like the auto-upgrade to Catalina broke the (unrelated) W3CDTF parser tests. I do not have a Mac to test :(. Should i add an issue for that?

Minoru commented:
> That comment is no longer relevant as source_count was deleted :)

Oh, you're totally right; sorry for the noise.

I'll go fix the Catalina problem, and get back here to review.

Minoru commented:
Catalina build is fixed; you may rebase onto master, expand the sources test, and the build should be fine :+1:

iglosiggio commented:
> Catalina build is fixed; you may rebase onto master, expand the sources test, and the build should be fine

I've rebased and added the test you requested. While doing that i've encountered a minor nitpick.

Media elements (`<video>` and `<audio>`) w/o valid sources do not appear on the rendered text but bump `video_count`/`audio_count` nonetheless. I can add the `source_count` variable again to "rollback" the count increment if there wasn't any valid source. I really don't think that adding that code has value.

Minoru commented:
Thank you for the rebase!

> I really don't think that adding that code has value.

I can't agree: the users will see the discrepancy and wonder where are the "missing" numbers. Even if it's not our fault that the \<source> is missing or invalid, we'll be the ones asked about it :) So in my opinion it makes sense to expend some effort on fixing that.

Luckily, it can be done without adding new code, just shuffling things around:

```c++
try {
    const std::string videourl = xpp.get_attribute_value("src");
    video_count++;
    add_media_link(curline, links, url, videourl,
        video_count, LinkType::VIDEO);
} catch (const std::invalid_argument&) {
    // It's not the end of the world if "src" attribute is invalid.
}
inside_video = true;
```

I.e. only increment the counter if `src` is valid. Ditto for audio and source tags.

What do you think of that?

iglosiggio commented:
>
> I.e. only increment the counter if src is valid. Ditto for audio and
> source tags.
>
> What do you think of that?.
>
The problem is that a video/audio doesn't have to have a valid src
attribute in order to be valid. Inside the tag multiple source tags can
provide different alternatives for the video (different qualities,
different encodings, etc).

To support that we need to add some kind of flag "had_valid_,sources"  and
decrement it after the closing (real or implicit) tag.

It's not a lot of code, if you really want it I can add an small commit
with the corresponding test.

Minoru commented:
Ah, right. Bummer. Please re-introduce the counters then. Having a test for that would be awesome as well.

iglosiggio commented:
Done! Do you think that's enough?

Minoru commented:
Yup, looks great @iglosiggio. Thank you very, very much for this PR and sticking with all my comments and requests for changes!
